### ISSUE TITLE
Bug: using with typescript v5.5 slows down linting by 2.3x

### ISSUE THREAD

**Issue by walkerdb** [typescript-eslint/typescript-eslint#9474]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

Under typescript 5.5 we're seeing typescript-eslint runs take 2.3x longer in our large monorepo than under any previous version of typescript. The slowdown applies to every typescript-eslint build we've tested, including the v8 rc. 

We narrowed this down to a specific typescript change by using `git bisect` in the typescript repo and running our typescript-eslint rules with local typescript builds: https://github.com/microsoft/TypeScript/pull/58045.

All builds prior to that commit run in expected time, all builds after run 2.3x slower. In our case that means 3m45s for ts <=5.4 vs ~8+ min for ts 5.5, with no changes other than the ts version.

I'll also be putting up an issue in the typescript repo since the slowdown seems to be coming on their end, but it seems worth raising here as well.

We can't share the private monorepo but I was able to reproduce a smaller 1.3x slowdown with the same setup in the sentry monorepo, chosen just because it's a large public monorepo that uses typescript-eslint. Our repo uses quite complex types that I'm assuming are somewhat of a pathological case for wherever the slowdown is coming from, which may explain the difference in slowdown size.

### Reproduction Repository Link

https://github.com/walkerdb/sentry/pull/1

### Repro Steps

1. after setting up deps, run `time yarn lint:js` in the ts 5.5 branch, observe the runtime
2. check out the linked ts 5.4 branch, install deps, and run the same command. It should be noticeably and consistently faster

I've put more detailed repro notes in the linked PR here: https://github.com/walkerdb/sentry/pull/1

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `rc-v8` |
| `@typescript-eslint/parser`            | `rc-v8` |
| `@typescript-eslint/scope-manager`     | `rc-v8` |
| `@typescript-eslint/typescript-estree` | `rc-v8` |
| `@typescript-eslint/type-utils`        | `rc-v8` |
| `@typescript-eslint/utils`             | `rc-v8` |
| `TypeScript`                           | `5.5.3` |
| `ESLint`                               | `8.57` |
| `node`                                 | `18.18` |

the same slowdown applies to typescript-eslint v7 and v6.

**Comment by JoshuaKGoldberg** [comment]
What a fantastic reproduction, thanks @walkerdb! ❤️‍🔥

Marking as blocked on an external issue to keep this out of our actionable queue. I'll revisit regularly to see how things are going.

**Comment by walkerdb** [comment]
A few perf traces we've run via [0x](https://github.com/davidmarkclements/0x) that show the ts slowdown. The red boxes are the time taken by the `walk` function. Before ts 5.5 they're fast, after 5.5 they're much slower, all from typescript itself.

| ts 5.4 | ts 5.5 |
| --- | --- |
| <img width="1497" alt="Screenshot 2024-07-01 at 5 48 38 PM" src="https://github.com/typescript-eslint/typescript-eslint/assets/7026330/86aa1f4d-5101-4c34-81d5-e81a92ff463b"> | <img width="1732" alt="Screenshot 2024-07-01 at 6 04 48 PM" src="https://github.com/typescript-eslint/typescript-eslint/assets/7026330/6e61a6c3-f55f-44e3-a5be-24279e00f595"> |

**Comment by walkerdb** [comment]
created a typescript issue here: https://github.com/microsoft/TypeScript/issues/59101

**Comment by maschwenk** [comment]
❤️ thank you for filing! I had the same issue and was fixed by the PR mentioned in https://github.com/microsoft/TypeScript/issues/59101

using:

```
"typescript": "npm:@typescript-deploys/pr-build@5.6.0-pr-59154-2",
```

**Comment by DanielRosenwasser** [comment]
TypeScript 5.5.4 should contain the fix - thanks for reporting everyone!

**Comment by JoshuaKGoldberg** [comment]
Thanks for hopping on this so quickly @DanielRosenwasser, we really appreciate it! ❤️

**Comment by JoshuaKGoldberg** [comment]
TypeScript 5.5.4 is released and TypeScript 5.6 is in beta. Closing this one out. If anybody has more performance issues, please file new issue(s) with full reproduction(s). Cheers all!
